### PR TITLE
Further ITK MPL cleanup (IsSame etc)

### PR DESCRIPTION
**Pull request by dzenanz**
## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [x] Added test (or behavior not changed)
- [x] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

### CODE REVIEW

File: Modules/Core/Common/include/itkInPlaceImageFilter.h [hunk 1]
```diff
@@ -147,7 +147,15 @@ class ITK_TEMPLATE_EXPORT InPlaceImageFilter : public ImageToImageFilter<TInputI
   void
   AllocateOutputs() override
   {
-    this->InternalAllocateOutputs(IsSame<TInputImage, TOutputImage>());
+    if (std::is_same<TInputImage, TOutputImage>::value)
```
Comment (N-Dekker) [inline Modules/Core/Common/include/itkInPlaceImageFilter.h:150]: Looks like a candidate for C++17 compile-time evaluated `if constexpr` 😸

File: Modules/Core/Common/test/itkMetaProgrammingLibraryTest.cxx [hunk 2]
```diff
@@ -19,77 +19,72 @@
...
-  itkStaticAssert((OrC<false, false, true>::Value == true), "Unit test failed");
-  itkStaticAssert((OrC<true, false, false>::Value == true), "Unit test failed");
-  itkStaticAssert((OrC<false, true, false>::Value == true), "Unit test failed");
-  itkStaticAssert((OrC<false, false, false>::Value == false), "Unit test failed");
-
-  itkStaticAssert((OrC<true, true>::Value == true), "Unit test failed");
-  itkStaticAssert((OrC<true, false>::Value == true), "Unit test failed");
-  itkStaticAssert((OrC<false, true>::Value == true), "Unit test failed");
-  itkStaticAssert((OrC<false, false>::Value == false), "Unit test failed");
+  static_assert((OrC<true, true, true>::Value == true), "Unit test failed");
```
Comment (N-Dekker) [inline Modules/Core/Common/test/itkMetaProgrammingLibraryTest.cxx]: Note that the extra pair of parentheses were necessary for the original `itkStaticAssert` macro calls, but are not needed for the static_assert's.

Comment (N-Dekker) [reply Modules/Core/Common/test/itkMetaProgrammingLibraryTest.cxx]: Ah, I see now, you removed those parentheses already!

File: Modules/Core/Common/include/itkInPlaceImageFilter.h (hunk 1 above)
Comment (dzenanz) [reply Modules/Core/Common/include/itkInPlaceImageFilter.h:150]: Even today, constant propagation should take care of optimization. If you have a TODO list of things for C++17, you could add this to it 😄

**Review by N-Dekker — APPROVED** [review]
Thanks @dzenanz ! 👍 👍 👍

**Pull request merged**